recirq/otoc/loschmidt/tilted_square_lattice/analysis.py

```diff
@@ -0,0 +1,172 @@
+    to group by, we group by all columns except the named columns. This can make analysis
+    more robust: if you add a new parameter it will automatically be included in the
+    groupby clauses without modifying the plotting code.
+
+    This method also does "SQL-style" groupby by setting `as_index=False` (preventing
+    row MultiIndices) and will flatten column MultiIndices to `_` delimited flat column names.
+
+    Args:
+        df: The dataframe
+        y_cols: Columnn not to group by. These will be aggregated over.
+        agg_func: The aggregation function
+
+    Returns:
+        The dataframe and list of columns used in the groupby.
+    """
+    groupby_cols = list(df.columns)
+    for yc in y_cols:
+        try:
+            groupby_cols.remove(yc)
+        except ValueError as e:
+            raise ValueError(f"{yc} not in columns {groupby_cols}.")
+
+    agged = df.groupby(groupby_cols, as_index=False).agg(agg_func)
+
+    if isinstance(agged.columns, pd.MultiIndex):
+        agged.columns = ['_'.join(x for x in column_tuple if x)
+                         for column_tuple in agged.columns]
+
+    return agged, groupby_cols
+
+
+def _results_to_dataframe(
+        results: ExecutableGroupResult,
+        func: Callable[[ExecutableResult, QuantumRuntimeConfiguration, SharedRuntimeInfo], Dict]
+) -> pd.DataFrame:
+    """Call a function on each result in an `ExecutableGroupResult` to construct a DataFrame."""
+    return pd.DataFrame([func(result, results.runtime_configuration, results.shared_runtime_info)
+                         for result in results.executable_results])
+
+
+def loschmidt_results_to_dataframe(results: ExecutableGroupResult) -> pd.DataFrame:
+    """Process an `ExecutableGroupResult`.
+
+    This function performs the data analysis using `to_ground_state_prob` and
+    extracts the most relevant quantities for a dataframe suitable for
+    grouping, selecting, and plotting.
+    """
+
+    def _to_record(result: ExecutableResult,
+                   rt_config: QuantumRuntimeConfiguration,
+                   shared_rt_info: SharedRuntimeInfo) -> Dict:
+        success_prob = to_ground_state_prob(result.raw_data)
+        spec = cast(TiltedSquareLatticeLoschmidtSpec, result.spec)
+
+        return {
+            'run_id': shared_rt_info.run_id,
+            'width': spec.topology.width,
+            'height': spec.topology.height,
+            'n_qubits': spec.topology.n_nodes,
+            'macrocycle_depth': spec.macrocycle_depth,
+            'q_area': (spec.macrocycle_depth * 4 * 2) * spec.topology.n_nodes,
+            'instance_i': spec.instance_i,
+            'n_repetitions': spec.n_repetitions,
+            'success_probability': success_prob,
+            'processor_id': rt_config.processor_record.processor_id,
+        }
+
+    return _results_to_dataframe(results, _to_record)
+
+
+def fit_vs_macrocycle_depth(df):
```

Comment:
Optional: Should we expose any of these functions in an __init__ file?

Reply:
maybe. It's easier to "promote" things to the init file than back off from it. For future algo benchmarks there could be e.g. two flavours of analysis in two different analysis modules and having the qualified name in notebooks could help with naming functions